File: week-04-fastapi/app/routers/tags.py
from fastapi import APIRouter, HTTPException
from bson import ObjectId
from app.database import get_db
from app.schemas import TagCreate, TagUpdate, TagResponse

router = APIRouter(prefix="/tags", tags=["tags"])


@router.get("", response_model=list[TagResponse])
def list_tags():
    db = get_db()
    tags = db.tags.find().sort("name", 1)
    return [TagResponse(**t) for t in tags]


@router.get("/{tag_id}", response_model=TagResponse)
def get_tag(tag_id: str):
    if not ObjectId.is_valid(tag_id):
        raise HTTPException(400, "Invalid tag ID")
    db = get_db()
    tag = db.tags.find_one({"_id": ObjectId(tag_id)})
    if not tag:
        raise HTTPException(404, "Tag not found")
    return TagResponse(**tag)


@router.post("", response_model=TagResponse, status_code=201)
def create_tag(body: TagCreate):
    db = get_db()
    result = db.tags.insert_one(body.model_dump())
    tag = db.tags.find_one({"_id": result.inserted_id})
    return TagResponse(**tag)


@router.put("/{tag_id}", response_model=TagResponse)
def update_tag(tag_id: str, body: TagUpdate):
    if not ObjectId.is_valid(tag_id):
        raise HTTPException(400, "Invalid tag ID")
    db = get_db()
    result = db.tags.update_one(
        {"_id": ObjectId(tag_id)},
        {"$set": body.model_dump()}
    )
    if result.matched_count == 0:
        raise HTTPException(404, "Tag not found")
    tag = db.tags.find_one({"_id": ObjectId(tag_id)})
    return TagResponse(**tag)


@router.delete("/{tag_id}", status_code=204)
def delete_tag(tag_id: str):
    if not ObjectId.is_valid(tag_id):
        raise HTTPException(400, "Invalid tag ID")
    db = get_db()
    result = db.tags.delete_one({"_id": ObjectId(tag_id)})
    if result.deleted_count == 0:
        raise HTTPException(404, "Tag not found")

File: week-05-express/src/routes/tags.js
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../db.js";

const router = Router();

router.get("/", async (req, res) => {
  const db = getDb();
  const tags = await db.collection("tags").find().sort({ name: 1 }).toArray();
  res.json(tags.map(t => ({ id: t._id.toString(), name: t.name })));
});

router.get("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) {
    return res.status(400).json({ detail: "Invalid tag ID" });
  }
  const db = getDb();
  const tag = await db.collection("tags").findOne({ _id: new ObjectId(req.params.id) });
  if (!tag) return res.status(404).json({ detail: "Tag not found" });
  res.json({ id: tag._id.toString(), name: tag.name });
});

router.post("/", async (req, res) => {
  const db = getDb();
  const result = await db.collection("tags").insertOne({ name: req.body.name });
  const tag = await db.collection("tags").findOne({ _id: result.insertedId });
  res.status(201).json({ id: tag._id.toString(), name: tag.name });
});

router.put("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) {
    return res.status(400).json({ detail: "Invalid tag ID" });
  }
  const db = getDb();
  const result = await db.collection("tags").updateOne(
    { _id: new ObjectId(req.params.id) },
    { $set: { name: req.body.name } }
  );
  if (result.matchedCount === 0) return res.status(404).json({ detail: "Tag not found" });
  const tag = await db.collection("tags").findOne({ _id: new ObjectId(req.params.id) });
  res.json({ id: tag._id.toString(), name: tag.name });
});

router.delete("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) {
    return res.status(400).json({ detail: "Invalid tag ID" });
  }
  const db = getDb();
  const result = await db.collection("tags").deleteOne({ _id: new ObjectId(req.params.id) });
  if (result.deletedCount === 0) return res.status(404).json({ detail: "Tag not found" });
  res.status(204).send();
});

export default router;
